example/lfo.cpp

```diff
@@ -0,0 +1,193 @@
+} dotted_quarter_note;
+inline constexpr struct quarter_note_triplet final : named_unit<"qt", mag_ratio<1, 3> * half_note> {
+} quarter_note_triplet;
+inline constexpr struct sixteenth_note final : named_unit<"16th", mag_ratio<1, 2> * eigth_note> {
+} sixteenth_note;
+inline constexpr struct dotted_eigth_note final : named_unit<"8th.", mag<3> * sixteenth_note> {
+} dotted_eigth_note;
+
+inline constexpr struct beats_per_minute final : named_unit<"bpm", quarter_note / non_si::minute> {
+} beats_per_minute;
+
+inline constexpr auto smpl = sample;
+
+inline constexpr auto pcm = sample_value;
+
+// inline constexpr auto n_wd = 3 * half_note;
+inline constexpr auto n_w = whole_note;
+// inline constexpr auto n_hd = dotted_half_note;
+// inline constexpr auto n_h = half_note;
+// inline constexpr auto n_qd = dotted_quarter_note;
+// inline constexpr auto n_q = quarter_note;
+// inline constexpr auto n_qt = quarter_note_triplet;
+// inline constexpr auto n_8thd = dotted_eigth_note;
+// inline constexpr auto n_8th = eigth_note;
+// inline constexpr auto n_16th = sixteenth_note;
+
+// inline constexpr auto bpm = beats_per_minute;
+
+//! Typesafe version of music application playback engine state
+struct musical_context {
+  quantity<sample_rate[si::hertz], float> sample_rate;
+  quantity<beats_per_minute, float> tempo;
+};
+
+//! Typesafe wrapper around API for host application musical context
+musical_context get_musical_context()
+{
+  auto context = third_party::get_musical_context();
+  return musical_context{.sample_rate = context.sample_rate * si::hertz, .tempo = context.tempo * beats_per_minute};
+}
+
+class sine_wave_osc {
+public:
+  sine_wave_osc(const musical_context& context, QuantityOf<isq::frequency> auto freq) :
+      m_context{context}, m_frequency{freq}
+  {
+    std::cout << std::format("Created LFO with starting frequency {} ({}) for sample rate {} at tempo {}\n", freq,
+                             m_frequency, context.sample_rate, context.tempo);
+  }
+
+  quantity<si::hertz, float> get_frequency() const { return m_frequency; }
+
+  void set_frequency(QuantityOf<isq::frequency> auto freq)
+  {
+    m_frequency = freq;
+    std::cout << std::format("Setting frequency to {} ({})\n", freq, m_frequency);
+  }
+
+  void set_period(QuantityOf<isq::time> auto period)
+  {
+    m_frequency = 1.f / period;
+    std::cout << std::format("Setting period to {} (i.e. frequency to {})\n", period, m_frequency);
+  }
+
+  quantity<sample_value, float> operator()()
+  {
+    auto out = sin(m_phase);
+    m_phase += step;
+    return out;
+  }
+
+private:
+  void update_step() { m_step = (angular::revolution * m_frequency) / m_context.sample_rate; }
+
+  quantity<sample_value> sin(quantity_point<angular::radian, float>)
+  {
+    return std::sin(m_phase.numerical_value_in(angular::radian)) * pcm;
+  }
+
+  musical_context m_context;
+  quantity<si::hertz, float> m_frequency;
+  quantity_point<angular::radian, float> m_phase{0.f};
```

Comment:
SI also has radian but does not have revolutions. I am not sure which one is the better use case considering that you use SI already.

Reply:
I'm not sure I have a strong opinion. I went with this both because of `angular::revolutions` and having a base dimension `angle` seems more natural than `angle` being `m / m`.

If you think it would be better for the example to use `si::radian` and `2π * si::radian` instead of `angle::revolutions`.

I can also poll my colleagues (who are not on holiday) tomorrow about this to get their thoughts.

Reply:
This was just an FYI so you can use the best tool for the job. You know better what is needed than me 😉